### ISSUE TITLE
If there is no buildout and no eggs, morepath.autosetup fails

### ISSUE THREAD

**Issue by href** [morepath/morepath#319]
I developed a kind of application server for my project that loads a configuration file and then starts a Morepath application. This works great in my development buildout environment, but it utterly fails when I deploy using pip.

Morepath tries to scan *all* my packages in this case (not just the ones depending on Morepath). This fails with various scanning errors, as not all packages lend themselves to be scanned.

What's interesting is that this only fails if there are no eggs. If I install my package using `pip install --egg` it works, at least under Python 3 (pip under Python 2 fails to actually create eggs for most packages).

I can work around the problem by either installing using `pip install --egg` under Python 3, or to not use morepath.autosetup. But I like morepath.autosetup and it would be great if it got more robust.

I don't really have a test case for this unfortunately, because I don't quiet see how to do that here. You can however reproduce the error by installing my package. Unfortunately this requires having Postgres and Memcached installed. If that is given, run this to reproduce the error:

```
brew install libmemcached
pip install onegov.town

curl https://raw.githubusercontent.com/OneGov/dev/master/onegov.yml.example > onegov.yml
onegov-server
```

The error seems to stem from the `morepath_packages` method. If no eggs are installed, the paths it finds are as follows:

```
/Users/denis/.virtualenvs/town-3-default/lib/python3.4/site-packages
/Users/denis/.virtualenvs/town-3-default/lib/python3.4/site-packages
... and so on ..
```
Though fixing those paths doesn't really help. I tried various small fixes, but I'm a bit confused how `morepath_packages` is supposed to work. It's function should basically be to return all modules that depend on Morepath, right? It seems like there should be a better, cleaner way. But then this is Python packaging, so I'm assuming this is messy :)

Anyway, I'm still investigating this, any hints would be appreciated.

**Comment by href** [comment]
I pushed a branch with a much simplified approach:
https://github.com/morepath/morepath/tree/issue_319

This one works in all environments, Python 2 (with, without eggs), Python3 (with, without eggs), Buildout. The only thing that doesn't work is loading the namespace package fixtures.

That might just be me, but it seems weird to me that the namespace fixtures are named `ns` and `ns2`, when I would expect them to be named `ns.real` and `ns2.real`. I assumed doing namespace packages meant naming them with the namespace included, but I can't find any canon about this. But in my years of working with Zope/Plone I haven't seen any packages that run under a namespace, without having that namespace prefixed to the package name.

In fact, I had a hard time importing those packages apart from `from ns import real`. Is this really something that needs to be supported?

Also am I being too naive with my `morepath_packages` version? What am I missing?

**Comment by faassen** [comment]
On Apr 30, 2015 11:25, "Denis Krienbühl" <notifications@github.com> wrote:
>

> That might just be me, but it seems weird to me that the namespace
fixtures are named ns and ns2, when I would expect them to be named ns.real
and ns2.real. I assumed doing namespace packages meant naming them with the
namespace included, but I can't find any canon about this. But in my years
of working with Zope/Plone I haven't seen any packages that run under a
namespace, without having that namespace prefixed to the package name.

Real quick, not much time now:

There is absolutely no requirement that project names in setup.py match
namespaces or module names. So if you rely on this you can't.

Also: the tox tests test this behavior and do not rely on buildout at all.
Are they installing things as eggs? I think tox uses pip.

Regards,

Martijn

**Comment by href** [comment]
Tox does indeed install everything as eggs, so those tests will not fail. I test all my packages in Tox anyway, so I would have seen these errors earlier.

> There is absolutely no requirement that project names in setup.py match namespaces or module names. So if you rely on this you can't.

My thinking is this: if only 1% of all namespace packages did this it might not be worth concentrating on this, given that the current implementation doesn't work with pip's defaults.

So far I've only gotten close to solving this case with the new approach using weird hacks that will probably fail in more exotic environments.

**Comment by href** [comment]
I tried fixing this for the fixture namespace packages, without success.

Barring other things my code is lacking, wouldn't it make sense to just say that Morepath packages with namespaces that are not included in the package name are not supported? There are no Morepath packages yet that violate such a rule and since this problem is bound to occur early in development of new Morepath packages, it seems like it's a rule that wouldn't hurt much.

**Comment by faassen** [comment]
Okay, I have a little bit of time to sit down an examine code now. :)

Some procedural feedback: I just noticed a branch is actually harder to understand for changes than a pull request. I don't know whether we can make pull requests from the project to itself, though! We need to find out a good way to discuss proposed code changes that works well with github.

Your code is vastly simpler than mine, which speaks highly for it. It avoids some nasty bug workarounds I had to do, too.

So what you're proposing is that the name of a distribution (in `setup.py`) should always be the package name, otherwise Morepath cannot automatically scan this package. Does this only apply to namespace packages or does it also apply to projects that contain a single non-namespaced package (or module) that is named differently from its name in `setup.py`? What about when there's only a capitalization difference?

Strictly speaking what you're doing in the modification is wrong: there is no relationship between the project name and any module or package name inside it. A project may even contain multiple packages. But practicality beats purity here, and my attempt at purity is hopelessly complicated and buggy compared to your more practical approach.

I am still hoping there is some unexplored setuptools API that lets us get at all the modules/packages listed in some magic way. But I certainly didn't find it last time as I ended up building this crufty code instead.

**Comment by href** [comment]
> Some procedural feedback: I just noticed a branch is actually harder to understand for changes than a pull request. I don't know whether we can make pull requests from the project to itself, though! We need to find out a good way to discuss proposed code changes that works well with github.

If you click on the green button when viewing a branch, you can look at a diff and create a pull request:
https://github.com/morepath/morepath/compare/issue_319

If you want I can always create a pull request up-front.

---

> So what you're proposing is that the name of a distribution (in `setup.py`) should always be the package name, otherwise Morepath cannot automatically scan this package. Does this only apply to namespace packages or does it also apply to projects that contain a single non-namespaced package (or module) that is named differently from its name in `setup.py`? What about when there's only a capitalization difference?

Well let me put it this way: I would love for the autoconfig to magically work for all kinds of packages. But I don't see a clean way to do it. I tried fixing the original function first to work if there are no eggs and at best I would've added another ugly hack on top of this. But I wasn't even able to do that.

So I tried rewriting the function and what I came up with works for everything I've ever worked on. Of course, because *my* packages are named after the module they contain. So it works for me ;)

It would work for everyone as long as the project_name is *exactly* like the module name that needs to be imported. That name would have to include the namespace and the case would matter.

I do realize though that not everyone does it that way. The question is if that is worth having a way more complicated way of doing it. 

Instead, how about we use entry points? I haven't used them myself for such a task, but the docs seem to indicate that they are good for exactly this use case:
http://pythonhosted.org/setuptools/pkg_resources.html#entry-points

So maybe we could try to import the package by using the project_name. If that doesn't work, we look for an entry_point that gives us more information. This could be a module or even a function that receives the config instance (so it could manually include other modules). I haven't fleshed out this idea in detail yet.

This would be easily testable and we could document this in detail. We could even show a warning if we can't import a package that depends on Morepath. It would be less magical, but the rules would be quite clear.

---

> I am still hoping there is some unexplored setuptools API that lets us get at all the modules/packages listed in some magic way. But I certainly didn't find it last time as I ended up building this crufty code instead.

Me too. I skimmed through too much setuptools/pkgutil/pkg_resources code by now :)

**Comment by faassen** [comment]
I needed to get back to this discussion.

Let's do the following:

* go with your approach. Only support automatic scanning if the package name is the same as the name in setup.py.

* document this clearly in the documentation on scanning.

* we also need a clear changelog entry that describes what might break and what you can do about it.

Then there's this step:

* support other cases using entry points. I've used them before but I was hoping we could avoid them and just exploit the morepath dependency relationship, but now that we cannot always do this, we should support the entry-point based method.

I'm fine with breaking this into two separate steps, doing the simplification of scanning and its documentation, and then making the entry point possible later.

**Comment by href** [comment]
I concur, let's do this in two steps. I'll find some time to document the new way in my branch. Once it is merged we can tackle entry points.

**Comment by href** [comment]
I drafted some language for the documentation:
https://github.com/morepath/morepath/commit/a35296baa65c75e8b605965456047542a1cc51f3

I'm not sure where else to put this and I don't know if this is enough. Please let me know what I can improve.

For a complete diff see https://github.com/morepath/morepath/compare/issue_319

**Comment by href** [comment]
I think we should also add a warning when the import fails. At that point we know that a package requires Morepath and we know what we could not import.

Can we re-raise the ImportError, amending it with our own message, pointing to the docs?

**Comment by faassen** [comment]
Feedback on the documentation:

  *Breaking change**. The ``morepath.autoconfig`` and ``morepath.autosetup``
  methods had to be rewritten.

Always say immediately why we make a change, in this case because it didn't work with pip (you do say this later).

We should give an actionable thing to do in the changelog as well, i.e. explain that you should either adjust the project name or load the stuff manually. Point out how you'd load something manually. (in imperative code that is executed before the server is started for now, unless we expect the entry point implementation lands soon. This requires using autoconfig instead of autosetup as you need the config object and do a manual commit, if I'm not mistaken)

in the API docs you say this:

  The package includes ``morepath`` in the ``install_requires`` list of
  the ``setup.py`` file.

but in fact it should still work if a project *independently* depends on Morepath, right?

The text in line 143 "If the name of the package differs from the name of its module" is very confusing, as "package" can be interpreted as a Python package (i.e. directory with __init__.py). Say something like 'if the setup.py name differs from the name of the distributed package or module...'

**Comment by faassen** [comment]
You are proposing to raise an import error, meaning that you'd see a traceback, not a warning, correct?

The question is what happens when we introduce entry points. We don't want an import error in thas case. Can we actually easily detect an entry point is for a particular package so we can skip doing the import error? More setuptools doc reading ahead. :)

**Comment by href** [comment]
Thanks for your feedback! I'll make the necessary changes once I have the chance.

> But in fact it should still work if a project independently depends on Morepath, right?

You mean a package depending on a package that depends on morepath? If so, it might be better to say "The package, or a dependency of the package includes `morepath` in the `install_requires` list of the `setup.py` file.

> You are proposing to raise an import error, meaning that you'd see a traceback, not a warning, correct?

Actually, an error would be raised anyway, if the name of the package doesn't match the name of its module. We currently try to import the project_name, which may or may not work.

I thought I could build on that and add a more specific message. A general ImportError might be very confusing.

> The question is what happens when we introduce entry points. We don't want an import error in thas case. Can we actually easily detect an entry point is for a particular package so we can skip doing the import error? More setuptools doc reading ahead. :)

Maybe we should have our own error in any case then. We could reuse that once we do entry points.
The entry points can easily be read form the distribution in `morepath_packages`. For example: `dist.get_entry_map()`. So I think we're quite flexible there.

**Comment by faassen** [comment]
Yes, "if the package or a dependency of that package includes morepath" would work.

If you don't see problems implementing the entry point system so that we can not do the auto-import at all when the entry point is there, I'm good with this.

**Comment by href** [comment]
Okay. I don't think we'll encounter any problems with entry points. But to be extra certain we can make sure to include this change as well as entry points in the same release. I still would tackle it as two separate issues though, since we'll have to discuss entry points more in-depth.

**Comment by faassen** [comment]
Okay! No need to go by me again for the additional doc changes. For the code changes a custom exception that we can't import the package even though it depends on Morepath would be good. For the entry point change we can then adjust this error to give a message about what entry point to add to setup.py.

**Comment by faassen** [comment]
(well, I guess we can't give the full entry point details as the whole point is we don't know the Python package name, but we can give instructions inline at least)

**Comment by href** [comment]
I finally got around to implementing the entry points. See https://github.com/morepath/morepath/compare/entry_points for an overview of all changes.

II kept the docs to a minimum, since I want to rewrite those for #327. It's getting way to cramped in the API docs. I'll add those docs to the entry_points branch as well. The idea is to close both this and #327 once the branch is merged.

Please let me know if you have any feedback. I'm a bit unsure about naming the entry point item `autoimport`. I also decided against supporting more than one autoimport item. I can't really think of a good use case for it atm.

Should there ever be a good use case it should be no problem to extend the current functionality in a backwards-compatible way.

**Comment by href** [comment]
Oh, and #326 fits in here as well.

**Comment by faassen** [comment]
Looks good. I think instead of 'autoimport' we can use 'scan'; that seems more consistent with Morepath naming.

**Comment by href** [comment]
I added a first draft of the new docs section and I renamed `autoimport` to `scan`. Let me know what you think. I haven't added any links to it yet outside the TOC.

**Comment by faassen** [comment]
Thanks for this! I've added quite a few comments to the draft now. :)

**Comment by href** [comment]
Thank for your suggestions, they really help :+1:

**Comment by href** [comment]
I hope I captured your feedback well enough. It's a lot of changes, so I don't think there's too much of a point if I go into every point separately. That'll get too complicated :)

Further feedback appreciated.

**Comment by href** [comment]
FYI, there were some issues in the docs that I just fixed. I guess it was getting a bit late for me yesterday ;)

**Comment by faassen** [comment]
I gave a lot of feedback again, some of it seems to be about the same I said before. If you prefer me to edit the text then just get it into trunk and give me a heads up, and I can refer her to my feedback again. I realize I'm being particular about getting the text the way I want to and I'm playing nitpicky editor. :)

**Comment by href** [comment]
I think you got a bit confused with the revisions. You've edited the same text again, the redacted version already implements a lot of your suggestions. To be honest though, I couldn't find a view with the latest version where you can add comments. I think we would have to merge this all into one first.

Maybe it's best I merge this for now and you have a look at it again? If you want you can edit yourself, or I can do it. I'm fine either way - I don't feel like this is my special child that only I can touch ;)

**Comment by faassen** [comment]
I was already wondering, it seemed so familiar. I just clicked on the last reference in this issue but that is obviously wrong. I wonder why the old comments didn't show up, that'd have clued me in..

**Comment by faassen** [comment]
I suggest you just merge it now and I can take a look at it.

**Comment by href** [comment]
Done, sorry it got so confusing.

**Comment by faassen** [comment]
It's not your fault, I should've looked better. Anyway, I took the document and edited it a bit. Lots of small tweaks and a new sidebar about scanning versus activation to explain scanning is a safe operation. I also linked to the API docs a bit more.

Thank you very much for this document, I hope this will help newcomers find their way better! I also linked it from the quickstart doc for that reason! I've rebuilt the latest docs on the doc website so people can see it.

I think we can close this issue now, right? I'll leave it up to you to do the honors. :)
